app/routes.py

```diff
@@ -1,2 +1,238 @@
+        return jsonify(
+                {"details": "Invalid data"}
+                ), 400
+    else: 
+        task = Task(title = request_body["title"],
+            description = request_body["description"],
+            completed_at = request_body["completed_at"])
+
+    db.session.add(task)
+    db.session.commit()
+
+    return {
+            "task": task.to_json()
+        }, 201
+
+@tasks_bp.route("", methods=["GET"])
+def get_all_tasks():
+
+#checks wave_2 tests 
+#sort title by asc and desc order
+    sort_by_title = request.args.get("sort")
+    if sort_by_title == "asc":
+        tasks = Task.query.order_by(Task.title.asc())
+    elif sort_by_title == "desc":
+        tasks = Task.query.order_by(Task.title.desc())
+    else:
+        tasks = Task.query.all()
+
+    tasks_response = []
+    for task in tasks:
+        tasks_response.append(task.to_json())
+    return jsonify(tasks_response)
+
+
+#GET PUT DELETE 
+@tasks_bp.route("/<task_id>", methods=["GET", "PUT", "DELETE"])
+def handle_task(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return make_response("", 404)
+
+    elif request.method == "GET":
+        return make_response({"task": task.to_json()}, 200)
+
+    elif request.method == "PUT":
+        form_data = request.get_json()
+
+        task.title = form_data["title"]
+        task.description = form_data["description"]
+        task.completed_at = form_data["completed_at"]
+
+        db.session.commit()
+
+        return make_response(
+            {"task": task.to_json()
+        }, 200)
+
+    elif request.method == "DELETE":
+        db.session.delete(task)
+        db.session.commit()
+        return make_response({
+            "details": f"Task {task.task_id} \"{task.title}\" successfully deleted"
+            })
+
+#wave_3 
+#creating custom endpoints with mark_complete: True or False
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"]) #update
+def mark_complete(task_id):
+    task = Task.query.get(task_id)
+
+    if not task:
+        return ("", 404)
+
+    task.completed_at = datetime.now()
+
+    db.session.commit()
+
+#wave_4
+#Slack API - sends messages to slack channel / uses message_to_slack helper function
+    message_to_slack("#task-notifications", f"Someone just completed the task {task.title}")
+
+    return {
+        "task": task.to_json()
+    }, 200
+
+#creating custom endpoints with mark_incomplete: True or False
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"]) #update
+def mark_incomplete(task_id):
+    task = Task.query.get(task_id)
+    if not task:
+        return ("", 404)
+
+    if task.completed_at != None:
+        task.completed_at = None
+    
+    db.session.commit()
+
+    return {
+        "task": task.to_json()
+        }, 200
+
+#checks wave_4 
+#created a function to send API message to slack 
+def message_to_slack(channel, message):
```

Comment:
Love that this is its own function! I think it makes the rest of the code easier to read!